### ISSUE TITLE
Flushing buffer

### ISSUE THREAD

**Issue by vanderlee** [reactphp/react#358]
I'm working on STARTTLS functionality for an SMTP server (according RFC3207)

This means that after the server receives a `STARTTLS` message from the client, it should send back a message `220 Ready to start TLS` and start accepting TLS negotiation using `stream_socket_enable_crypto()`.

The client would then receive the `220 Ready to start TLS` and start TLS negotiation.

The problem is that the server needs to write a message to the socket _and_ open the socket to accept TLS negotiation. When I call `stream_socket_enable_crypto()`, the `220 ...` message is not sent to the client, so the client doesn't start the TLS negotiation.

Is there a way to flush the data to the client and immediately call `stream_socket_enable_crypto()` afterwards?

**Comment by vanderlee** [comment]
Found it. You can subscribe to the `end` or `close` events on the `\React\StreamStream` class, which will fire  after ensuring the buffer is flushed and just before closing the stream.

Not quite as nice as being able to explicitely flush the buffer, but workable for my particular situation.

**Comment by clue** [comment]
Hi @vanderlee, thanks for filing this issue!

It looks like your issue has been resolved and this can be closed?

Just for the reference, this is a common request and there are also some answers in https://github.com/reactphp/stream/pull/16#issuecomment-71862449 and https://github.com/reactphp/stream/issues/43.

I hope this helps :+1:

**Comment by vanderlee** [comment]
My issue is solved by using the `close` event to wait for flushing of the output (though it's not really a "clean" solution as it uses a mechanism intended for a different reason), so you can close this issue.

But at the very least the use case of switching to `stream_socket_enable_crypto()` as required by SMTP's `STARTTLS` could be the reason you were looking for in https://github.com/reactphp/stream/issues/43#issuecomment-211278124

As I understand it, the `STARTTLS` mechanism as used in SMTP is also used in a number of other protocols; https://en.wikipedia.org/wiki/Opportunistic_TLS

p.s. subscribing to `full-drain` won't work in this case, since the socket is closed before the function is triggered and `stream_socket_enable_crypto()` needs an open socket.

**Comment by clue** [comment]
I'm not sure I understand your particular code, as the `close` event will only be emitted when the stream is being closed, whereas your use case sounds like you want to reuse the same stream to enable encryption. Have you tried something like this?

```php
$encryption = new StreamEncryption(); // from SocketClient component
$stream->getBuffer()->on('full-drain', function () use ($stream, $encryption) {
    // buffer is now empty / flushed, let's enable encryption
   $encryption->enable($stream)->then(function () use ($stream) {
       // encryption is now established
       $stream->write("this is now encrypted");
   });
});
$stream->write("final plain text write to confirm STARTTLS");

**Comment by vanderlee** [comment]
I tried the code you suggested and it worked, but not  without changing React.

tl;dr: class `StreamEncryption` only allows client-side TLS negotiation.

Long version; here's the code I ended up using:
```
			$crypto_method = STREAM_CRYPTO_METHOD_TLS_SERVER;
			if (defined('STREAM_CRYPTO_METHOD_TLSv1_2_SERVER')) {
				$crypto_method |= STREAM_CRYPTO_METHOD_TLSv1_2_SERVER | STREAM_CRYPTO_METHOD_TLSv1_1_SERVER; // TLSv1.0 no longer considered safe to use.
			}
		
			stream_context_set_option($this->stream, 'ssl', 'local_cert', 'path/to/my_certificate.pem'); // because TLS server side requires a certificate (self-signed works).
			stream_context_set_option($this->stream, 'ssl', 'passphrase', ''); // If certificate has passphrase, otherwise just null.

			$encryption = new \React\SocketClient\StreamEncryption($this->loop);
			$encryption->method = $crypto_method;
			$stream->getBuffer()->on('full-drain', function () use ($stream, $encryption) {
				stream_set_blocking($stream->stream, true); // apparently more reliable to block during TLS negotiation
				$encryption->enable($stream);
				stream_set_blocking($stream->stream, false);
				// STARTTLS for SMTP must not send a confirmation message.
			});
			$stream->write("220 Ready to start TLS", true);
```
As you can see, my hacky fix to use the server-side TLS is by making the `$method` property of `StreamEncryption` public.
I'm sure there's a _much_ better way of doing this.
Perhaps checking whether the Stream is listening to a port to see if it's a server. I don't know how reliable this could be and whether some protocols require a role reversal.
Perhaps an optional extra parameter (i.e. `$server`) to the constructor (or `enable` method) of `StreamEncryption` (defaulting to `false` for backwards compatibility).

**Comment by clue** [comment]
Thanks for the confirmation and reporting this back! :+1: I believe this ticket is resolved and will close this for now, please report back it this persists :+1: 

I probably misread your OP, didn't realize you're talking about a server component here. Either way, glad you've got this sorted, as official TLS support on the server side is still in the works.
For future reference, see also https://github.com/reactphp/socket/issues/24

For the time being, keep in mind that making the stream blocking may mess with the loop if your clients are not under your control and don't immediately send a TLS ClientHello.
